src/lm_robust_helper.cpp

```diff
@@ -163,34 +289,70 @@ List lm_variance(const Eigen::Map<Eigen::MatrixXd>& X,
       bread = Kr(Eigen::MatrixXd::Identity(ny, ny), XtX_inv);
     }
 
-    if ( !clustered ) {
-      // Robust, no clusters
-
-      if (type == "HC2") {
+    Eigen::MatrixXd meatXtX_inv;
+    if ((se_type == "HC2") || (se_type == "HC3") || (se_type == "CR2")) {
+      if (X.cols() > r) {
+        meatXtX_inv = getMeatXtX(X, XtX_inv);
+        r_fe = meatXtX_inv.cols();
+      } else {
+        meatXtX_inv = XtX_inv;
+      }
+    }
+    // Rcout << "meatXtX_inv:" << std::endl << meatXtX_inv << std::endl;
 
-        for (int i = 0; i < n; i++) {
-          Eigen::VectorXd Xi = X.row(i);
-          temp_omega.row(i) = temp_omega.row(i) / (1.0 - (Xi.transpose() * XtX_inv * Xi));
-        }
+    if ( !clustered ) {
+      // Rcout << "temp_omega:" << std::endl << temp_omega << std::endl;
 
-      } else if (type == "HC3") {
+      if ((se_type == "HC2") || (se_type == "HC3")) {
 
         for (int i = 0; i < n; i++) {
-          Eigen::VectorXd Xi = X.row(i);
-          temp_omega.row(i) = temp_omega.row(i) / (std::pow(1.0 - Xi.transpose() * XtX_inv * Xi, 2));
+          Eigen::ArrayXd new_omega(ny);
+          Eigen::VectorXd Xi = X.leftCols(r_fe).row(i);
+          // Rcout << i << ":" << Xi << std::endl;
+
+          if (se_type == "HC2") {
+            new_omega = temp_omega.row(i) / (1.0 - (Xi.transpose() * meatXtX_inv * Xi));
+          } else if (se_type == "HC3") {
+            new_omega = temp_omega.row(i) / (std::pow(1.0 - Xi.transpose() * meatXtX_inv * Xi, 2));
+          }
+          // Rcout << i << ":" << new_omega << std::endl;
+          // Perfect fits cause instability, but we can place 0s for those
+          // observations and the rest of the estimation works
+          for (int j = 0; j < ny; j++) {
+            if (std::isnan(new_omega(j)) || std::isinf(new_omega(j))) {
+              new_omega(j) = 0.0;
+            }
+          }
```

Comment:
https://stackoverflow.com/questions/22927323/how-do-i-find-and-replace-all-non-finite-numbers-in-an-eigenarray-object

Reply:
Blessed.